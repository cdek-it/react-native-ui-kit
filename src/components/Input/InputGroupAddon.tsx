import type { Icon } from '@tabler/icons-react-native'
import { createElement, memo } from 'react'
import { Pressable, Text } from 'react-native'

import { makeStyles } from '../../utils/makeStyles'

export interface InputGroupAddonProps {
  /** Содержимое аддона инпут группы, текст или иконка из Tabler */
  content: string | Icon
  /** Расположение аддона слева или справа в группе */
  position: 'left' | 'right'
  /** Управление активностью аддона */
  disabled?: boolean
  /** Обработчик нажатия */
  onPress?: () => void
}

/**
 * Служебный компонент для группировки инпута
 * @link https://www.figma.com/design/4TYeki0MDLhfPGJstbIicf/UI-kit-PrimeFace-(DS)?node-id=484-5932&m=dev
 * @see InputGroup
 */
export const InputGroupAddon = memo<InputGroupAddonProps>(
  ({ content, onPress, position, disabled }) => {
    const styles = useStyles()

    return (
      <Pressable
        collapsable={false}
        disabled={disabled}
        style={[
          styles.container,
          styles[position],
          disabled && styles.disabled,
        ]}
        testID='InputGroupAddon_Pressable'
        onPress={onPress}
      >
        {typeof content === 'string' ? (
          <Text style={styles.text}>{content}</Text>
        ) : (
          createElement(content, {
            style: styles.icon,
            width: styles.icon.width,
            height: styles.icon.height,
          })
        )}
      </Pressable>
    )
  }
)

const useStyles = makeStyles(({ theme }) => ({
  container: {
    paddingVertical: theme.Form.InputText.inputPaddingTopBottom,
    paddingHorizontal: theme.Form.InputText.inputPaddingLeftRight,
    justifyContent: 'center',
    borderRadius: theme.General.borderRadius,
    borderWidth: 1,
    borderColor: theme.Form.InputText.inputBorderColor,
    backgroundColor: theme.Form.InputGroup.inputGroupBg,
  },
  left: {
    borderRightWidth: 0,
    borderTopRightRadius: 0,
    borderBottomRightRadius: 0,
  },
  right: {
    borderLeftWidth: 0,
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0,
  },
  disabled: {
    opacity: 0.6,
    backgroundColor: theme.Button.Disabled.disabledButtonBg,
  },
  text: {
    fontSize: 14,
    color: theme.Form.InputGroup.inputGroupTextColor,
    includeFontPadding: false,
    verticalAlign: 'middle',
  },
  icon: {
    width: 14,
    height: 14,
    color: theme.Form.InputGroup.inputGroupTextColor,
  },
}))
