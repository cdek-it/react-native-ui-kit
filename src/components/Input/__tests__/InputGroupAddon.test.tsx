import { IconUser } from '@tabler/icons-react-native'
import { fireEvent, render } from '@testing-library/react-native'

import { type InputGroupAddonProps, InputGroupAddon } from '../InputGroupAddon'

jest.mock('react-native/Libraries/Components/Pressable/Pressable', () => ({
  default: 'Pressable',
}))

describe('InputGroup component tests', () => {
  const snapshotCases: Array<[string, InputGroupAddonProps]> = [
    ['content: string, position: left', { content: 'text', position: 'left' }],
    [
      'content: icon, position: right',
      { content: IconUser, position: 'right' },
    ],
    ['disabled: true', { content: 'text', position: 'left', disabled: true }],
    ['disabled: false', { content: 'text', position: 'left', disabled: false }],
  ]

  test.each(snapshotCases)('%s', (_, props) => {
    const renderedInputText = render(<InputGroupAddon {...props} />)

    expect(renderedInputText.toJSON()).toMatchSnapshot()
  })

  test('should handle onPress', () => {
    const onPress = jest.fn()
    const { queryByTestId } = render(
      <InputGroupAddon content='text' position='left' onPress={onPress} />
    )
    const pressable = queryByTestId('InputGroupAddon_Pressable')

    fireEvent.press(pressable)

    expect(onPress).toHaveBeenCalled()
  })
})
