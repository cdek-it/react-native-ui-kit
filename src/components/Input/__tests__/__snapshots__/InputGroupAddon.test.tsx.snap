// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputGroup component tests content: icon, position: right 1`] = `
<Pressable
  collapsable={false}
  style={
    [
      {
        "backgroundColor": "#ffffff",
        "borderColor": "#dbdada",
        "borderRadius": 7,
        "borderWidth": 1,
        "justifyContent": "center",
        "paddingHorizontal": 10.5,
        "paddingVertical": 0,
      },
      {
        "borderBottomLeftRadius": 0,
        "borderLeftWidth": 0,
        "borderTopLeftRadius": 0,
      },
      undefined,
    ]
  }
  testID="InputGroupAddon_Pressable"
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={14}
    bbWidth={14}
    fill="none"
    focusable={false}
    height={14}
    meetOrSlice={0}
    minX={0}
    minY={0}
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "color": "rgba(0, 0, 0, 0.6000)",
          "height": 14,
          "width": 14,
        },
        {
          "flex": 0,
          "height": 14,
          "width": 14,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={14}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      color="rgba(0, 0, 0, 0.6000)"
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "type": 2,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2}
    >
      <RNSVGPath
        color="rgba(0, 0, 0, 0.6000)"
        d="M8 7a4 4 0 1 0 8 0a4 4 0 0 0 -8 0"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        color="rgba(0, 0, 0, 0.6000)"
        d="M6 21v-2a4 4 0 0 1 4 -4h4a4 4 0 0 1 4 4v2"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</Pressable>
`;

exports[`InputGroup component tests content: string, position: left 1`] = `
<Pressable
  collapsable={false}
  style={
    [
      {
        "backgroundColor": "#ffffff",
        "borderColor": "#dbdada",
        "borderRadius": 7,
        "borderWidth": 1,
        "justifyContent": "center",
        "paddingHorizontal": 10.5,
        "paddingVertical": 0,
      },
      {
        "borderBottomRightRadius": 0,
        "borderRightWidth": 0,
        "borderTopRightRadius": 0,
      },
      undefined,
    ]
  }
  testID="InputGroupAddon_Pressable"
>
  <Text
    style={
      {
        "color": "rgba(0, 0, 0, 0.6000)",
        "fontSize": 14,
        "includeFontPadding": false,
        "verticalAlign": "middle",
      }
    }
  >
    text
  </Text>
</Pressable>
`;

exports[`InputGroup component tests disabled: false 1`] = `
<Pressable
  collapsable={false}
  disabled={false}
  style={
    [
      {
        "backgroundColor": "#ffffff",
        "borderColor": "#dbdada",
        "borderRadius": 7,
        "borderWidth": 1,
        "justifyContent": "center",
        "paddingHorizontal": 10.5,
        "paddingVertical": 0,
      },
      {
        "borderBottomRightRadius": 0,
        "borderRightWidth": 0,
        "borderTopRightRadius": 0,
      },
      false,
    ]
  }
  testID="InputGroupAddon_Pressable"
>
  <Text
    style={
      {
        "color": "rgba(0, 0, 0, 0.6000)",
        "fontSize": 14,
        "includeFontPadding": false,
        "verticalAlign": "middle",
      }
    }
  >
    text
  </Text>
</Pressable>
`;

exports[`InputGroup component tests disabled: true 1`] = `
<Pressable
  collapsable={false}
  disabled={true}
  style={
    [
      {
        "backgroundColor": "#ffffff",
        "borderColor": "#dbdada",
        "borderRadius": 7,
        "borderWidth": 1,
        "justifyContent": "center",
        "paddingHorizontal": 10.5,
        "paddingVertical": 0,
      },
      {
        "borderBottomRightRadius": 0,
        "borderRightWidth": 0,
        "borderTopRightRadius": 0,
      },
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "opacity": 0.6,
      },
    ]
  }
  testID="InputGroupAddon_Pressable"
>
  <Text
    style={
      {
        "color": "rgba(0, 0, 0, 0.6000)",
        "fontSize": 14,
        "includeFontPadding": false,
        "verticalAlign": "middle",
      }
    }
  >
    text
  </Text>
</Pressable>
`;
